modules/hikari/src/test/scala/doobie/postgres/PGConcurrentSuite.scala

```diff
@@ -62,4 +62,34 @@ class PGConcurrentSuite extends munit.FunSuite {
     assertEquals(pollingStream.unsafeRunSync(), ())
   }
 
+  test("Connection returned before stream is drained") {
+    val xa = Transactor.fromDataSource[IO](
+      dataSource,
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(32))
+    )
+
+    val count = 100
+    val insert = for {
+      _ <- Stream.eval(sql"CREATE TABLE if not exists stream_cancel_test (i text)".update.run.transact(xa))
+      _ <- Stream.eval(sql"truncate table stream_cancel_test".update.run.transact(xa))
+      _ <- Stream.eval(sql"INSERT INTO stream_cancel_test values ('1')".update.run.transact(xa)).repeatN(count)
+    } yield ()
+
+    insert.compile.drain.unsafeRunSync()
+
+    val streamLargerBuffer = fr"select * from stream_cancel_test".query[Int].stream.transactBuffer(xa, 1024)
+      .evalMap(_ => fr"select 1".query[Int].unique.transact(xa))
+      .compile.count
+
+    assertEquals(streamLargerBuffer.unsafeRunSync(), count.toLong)
+
+    // if buffer is less than result set, it will be still block new connection since the result set is not drained
+    // use sleep to test the result set can be drained
+    val streamSmallerBuffer = fr"select * from stream_cancel_test".query[Int].stream.transactBuffer(xa, 50)
```

Comment:
Hm sorry I can't quite understand how this is testing that the connection is closed before the stream is completed :thinking: 

An alternative way I can think of to test this is to use some sort of atomic boolean and set it to true when the transactor finishes the transaction, and then you can assert that this was set to true before you process the last chunk of the stream.
```
  import doobie.util.transactor.Strategy
  import doobie.FC
  import java.util.concurrent.atomic.AtomicBoolean
  val hasClosed = new AtomicBoolean(false)
  xa.copy(
    strategy0 = Strategy.default.copy(
      always = Strategy.default.always.flatMap(_ => FC.delay(hasClosed.set(true)))
    )
  )
```

Reply:
Fixed the test to assert if it's closed